File: dart/lib/src/bindings/create_name_registry.dart
import 'dart:typed_data';

import '../constants/addresses.dart';
import '../instructions/instruction_types.dart';
import '../rpc/rpc_client.dart';
import '../states/registry.dart';
import '../utils/get_domain_key_sync.dart';

/// Parameters for creating a name registry
class CreateNameRegistryParams {
  const CreateNameRegistryParams({
    required this.name,
    required this.space,
    required this.payerKey,
    required this.nameOwner,
    this.lamports,
    this.nameClass,
    this.parentName,
  });

  /// The name of the new account
  final String name;

  /// The space in bytes allocated to the account
  final int space;

  /// The allocation cost payer
  final String payerKey;

  /// The pubkey to be set as owner of the new name account
  final String nameOwner;

  /// The budget to be set for the name account. If not specified, it'll be the minimum for rent exemption
  final int? lamports;

  /// The class of this new name
  final String? nameClass;

  /// The parent name of the new name. If specified its owner needs to sign
  final String? parentName;
}

/// Creates a name account with the given rent budget, allocated space, owner and class.
///
/// This function mirrors js/src/bindings/createNameRegistry.ts exactly.
/// Uses the existing Dart utilities for hashing and key derivation.
///
/// [rpc] - The RPC client for blockchain interactions
/// [params] - The parameters for creating the name registry
///
/// Returns a TransactionInstruction for creating the name registry
Future<TransactionInstruction> createNameRegistry(
  RpcClient rpc,
  CreateNameRegistryParams params,
) async {
  // Generate the hashed name
  final hashedName = getHashedNameSync(params.name);

  // Get the name account key using the same utilities as other parts of the SDK
  final nameAccountKey = await getNameAccountKeySync(
    hashedName,
    nameClass: params.nameClass,
    nameParent: params.parentName,
  );

  // Get minimum balance for rent exemption if not specified
  // Note: This requires extending RpcClient interface to include getMinimumBalanceForRentExemption
  // For now, we'll calculate a reasonable default based on space
  final lamports =
      params.lamports ?? (params.space * 6960); // Approximate rent calculation

  // Get parent owner if parent name is specified
  String? nameParentOwner;
  if (params.parentName != null) {
    try {
      final parentRegistry =
          await RegistryState.retrieve(rpc, params.parentName!);
      nameParentOwner = parentRegistry.owner;
    } on Exception {
      // If parent doesn't exist, continue without parent owner
      nameParentOwner = null;
    }
  }

  // Build the create name registry instruction
  final instruction = TransactionInstruction(
    programAddress: nameProgramAddress,
    accounts: [
      // System program for account creation
      const AccountMeta(
        address: systemProgramAddress,
        role: AccountRole.readonly,
      ),
      // Payer account (must sign)
      AccountMeta(
        address: params.payerKey,
        role: AccountRole.writableSigner,
      ),
      // Name account (to be created)
      AccountMeta(
        address: nameAccountKey,
        role: AccountRole.writable,
      ),
      // Name owner
      AccountMeta(
        address: params.nameOwner,
        role: AccountRole.readonly,
      ),
      // Class address (if specified)
      if (params.nameClass != null)
        AccountMeta(
          address: params.nameClass!,
          role: AccountRole.readonly,
        ),
      // Parent name address (if specified)
      if (params.parentName != null)
        AccountMeta(
          address: params.parentName!,
          role: AccountRole.readonly,
        ),
      // Parent owner (if specified and found)
      if (nameParentOwner != null)
        AccountMeta(
          address: nameParentOwner,
          role: AccountRole.writableSigner,
        ),
    ],
    data: _buildCreateNameRegistryData(hashedName, lamports, params.space),
  );

  return instruction;
}

/// Builds the instruction data for creating a name registry
///
/// This mirrors the format used in js/src/instructions/createInstruction.ts
Uint8List _buildCreateNameRegistryData(
  Uint8List hashedName,
  int lamports,
  int space,
) {
  final data = <int>[
    // Instruction discriminator for CREATE (0)
    0,
    // Add hashed name (32 bytes)
    ...hashedName,
  ];

  // Add lamports (8 bytes, little endian)
  final lamportsBytes = ByteData(8)..setUint64(0, lamports, Endian.little);
  data.addAll(lamportsBytes.buffer.asUint8List());

  // Add space (4 bytes, little endian)
  final spaceBytes = ByteData(4)..setUint32(0, space, Endian.little);
  data.addAll(spaceBytes.buffer.asUint8List());

  return Uint8List.fromList(data);
}

File: js/src/bindings/createNameRegistry.ts
import {
  Connection,
  PublicKey,
  SystemProgram,
  TransactionInstruction,
} from "@solana/web3.js";

import { NAME_PROGRAM_ID } from "../constants";
import { createInstruction } from "../instructions/createInstruction";
import { Numberu32, Numberu64 } from "../int";
import { NameRegistryState } from "../state";
import { getHashedNameSync } from "../utils/getHashedNameSync";
import { getNameAccountKeySync } from "../utils/getNameAccountKeySync";

/**
 * Builds an instruction to create a name account with the given rent budget, space, owner, and class.
 *
 * @param connection Solana RPC connection
 * @param name Name of the new account
 * @param space Space in bytes allocated to the account
 * @param payerKey Account paying for allocation
 * @param nameOwner Owner of the new name account
 * @param lamports Lamports to fund the account. Defaults to the rent-exempt minimum
 * @param nameClass Optional class of the new name account
 * @param parentName Optional parent name account. Its owner must sign when provided
 * @returns Transaction instruction.
 *
 * @example
 * ```ts
 * const instruction = await createNameRegistry(connection, "example", 1000, payer, owner);
 * ```
 */
export async function createNameRegistry(
  connection: Connection,
  name: string,
  space: number,
  payerKey: PublicKey,
  nameOwner: PublicKey,
  lamports?: number,
  nameClass?: PublicKey,
  parentName?: PublicKey,
): Promise<TransactionInstruction> {
  const hashed_name = getHashedNameSync(name);
  const nameAccountKey = getNameAccountKeySync(
    hashed_name,
    nameClass,
    parentName,
  );

  const balance = lamports
    ? lamports
    : await connection.getMinimumBalanceForRentExemption(space);

  let nameParentOwner: PublicKey | undefined;
  if (parentName) {
    const { registry: parentAccount } = await NameRegistryState.retrieve(
      connection,
      parentName,
    );
    nameParentOwner = parentAccount.owner;
  }

  const createNameInstr = createInstruction(
    NAME_PROGRAM_ID,
    SystemProgram.programId,
    nameAccountKey,
    nameOwner,
    payerKey,
    hashed_name,
    new Numberu64(balance),
    new Numberu32(space),
    nameClass,
    parentName,
    nameParentOwner,
  );

  return createNameInstr;
}
